### ISSUE TITLE
Expose formatter to public or protected in CommonWriter

### ISSUE THREAD

**Issue by c0nstexpr** [touchlab/Kermit#360]
https://github.com/touchlab/Kermit/blob/8e20fa506771c5078b55c0a32ef93e2889aba10e/kermit-core/src/commonMain/kotlin/co/touchlab/kermit/CommonWriter.kt#L19
At now CommonWriter's formatter is private, which make it impossible to use custom formatter when inheriting writer. See below
```
class MyWriter() : CommonWriter(MyFormatter()) {
    override fun log(severity: Severity, message: String, tag: String, throwable: Throwable?) =
        Mylog.log(messageStringFormatter.formatMessage(severity, Tag(tag), Message(message)))
}
```

**Comment by findjigar** [comment]
It's possible to customize Message Formatting. Did you check out this docs? https://kermit.touchlab.co/docs/configuration/MESSAGE_FORMATTING

**Comment by c0nstexpr** [comment]
> It's possible to customize Message Formatting. Did you check out this docs? https://kermit.touchlab.co/docs/configuration/MESSAGE_FORMATTING

Yes, I know the current way to customize log writer. And I also successfully implement my own [writer](https://github.com/c0nstexpr/fishology/blob/c50e09605871bfdca05735e715e154e34501327e/core/src/client/kotlin/org/c0nstexpr/fishology/log/MCMessageWriter.kt#L12). 
But if we implement a class to inherit CommonWriter, providing customization by overriding base log method seems to be pointless. Because the inheritor cannot access the formatter.

**Comment by kpgalligan** [comment]
Sorry for the long delay. Not sure the original poster cares at this point, but I'm cleaning up the issues list. Lots of them perpetually stay in the "we'll look at this when we get some extra time", but that never happens on its own.

The idea of `MessageStringFormatter` was to provide a way to customize formatting without needing to write custom writers. If you were going to create a custom writer, you'd either be doing it because you wanted more control over formatting, or you'd need to take the `MessageStringFormatter` instance as a constructor arg and could just hold onto it in your custom code. But, I can see the counter argument.

I'm fairly reluctant to expand the API too much, but this might be worth it. I'd just want to think through it a bit.

**Comment by c0nstexpr** [comment]
> Sorry for the long delay. Not sure the original poster cares at this point, but I'm cleaning up the issues list. Lots of them perpetually stay in the "we'll look at this when we get some extra time", but that never happens on its own.
> 
> The idea of `MessageStringFormatter` was to provide a way to customize formatting without needing to write custom writers. If you were going to create a custom writer, you'd either be doing it because you wanted more control over formatting, or you'd need to take the `MessageStringFormatter` instance as a constructor arg and could just hold onto it in your custom code. But, I can see the counter argument.
> 
> I'm fairly reluctant to expand the API too much, but this might be worth it. I'd just want to think through it a bit.

Yes, it is all about api design. 
My first idea to implement my own logger, is to check **whether library has provide a good way for me to do it.**
Then I see the `CommonWriter` qualified by `open`. And I'm more confused by the `private` member in ctor.

Originally, I would write something like
```
class MyWriter() : CommonWriter(MyFormatter()) {
    override fun log(severity: Severity, message: String, tag: String, throwable: Throwable?) {
        client.msg(
            coloredText(
                levelFmt.getOrElse(severity) { severity.defaultFmt() },
                messageStringFormatter.formatMessage(severity, Tag(tag), Message(message)),
            ),
        )
    }
}
```
